# Perkenalan
Nama : Hillidatul Ilmi
Batch : 12

# Problems
a. Buatlah dua fungsi eksplisit masing-masing memuat perhitungan hambatan (R) dan daya (P) yang nantinya akan digunakan untuk perhitungan

In [87]:
# Eksplisit Perhitungan Hambatan
def eksplisit_hambatan (R,V):
    hambatan = R/V
    return(R/V)
    eksplisit_hambatan(8,2)

In [88]:
# Eksplisit Perhitungan Daya
def eksplisit_daya (R,I):
    daya = R*I
    return(R*I)
    eksplisit_daya(3,2)

b. Buat kolom baru dengan nama kolom R/P berisikan hasil perhitungan daya/hambatan dengan kriteria untuk Kapasitas >=2.5 hitung daya, Kapasitas <2.5 hitung nilai hambatan. (Hint: Gunakan for loop dan if statement untuk mengerjakan perintah nomor b, kamu bisa menggunakan list untuk menyimpan perhitungan sementara, setelahnya dapat dimasukkan ke kolom R/P).

In [89]:
import numpy as pd
import numpy as np

In [90]:
import pandas as pd
# Membuat Kolom baru
kolom = pd.DataFrame({'Ampere':[2,4,6,8,10],'cycle':[3,5,7,9,11]})
kolom

Ampere  cycle
0       2      3
1       4      5
2       6      7
3       8      9
4      10     11

In [91]:
# hasil perhitungan daya/hambatan dengan kriteria untuk Kapasitas >=2.5 hitung daya, 
# Kapasitas <2.5 hitung nilai hambatan.
nilai = 2

if 2.5< nilai >= 2.5 :
    elif nilai >=2.5:
    print(f'{nilai} Masuk Kriteria Kapasitas daya')
    else angka <2.5:
        print(f'{nilai} Masuk Kriteria Kapasitas hambatan')

SyntaxError: invalid syntax (1025764880.py, line 6)

c. Hitunglah rata-rata nilai kapasitas untuk arus = 1 Ampere dan cycle ganjil.

In [ ]:
pd.DataFrame().mean('Ampere')

# Assignment Rubrics

Keterangan kolom:

time(s): Detik ke- pengukuran
V_m: Voltage/tegangan yang terukur tiap waktu
Current (A): Kuat arus terukur tiap waktu
Cycle: Nomor siklus charge-discharge baterai
Cap: Kapasitas baterai tiap waktu

In [ ]:
import pandas as pd
import numpy as np

In [ ]:
# Data Loading	Mampu memuat data dengan Panda
d = pd.read_csv('NMC1-9.csv')
d

time(s)    V_m  Current (A)  Cycle       Cap
0            0  4.098        0.997      1  2.840817
1            1  4.096        0.998      1  2.840817
2            2  4.094        0.999      1  2.840817
3            3  4.093        0.998      1  2.840817
4            4  4.092        0.999      1  2.840817
...        ...    ...          ...    ...       ...
90977    90977  2.506        1.002      9  2.770180
90978    90978  2.504        1.002      9  2.770180
90979    90979  2.502        1.002      9  2.770180
90980    90980  2.500        1.002      9  2.770180
90981    90981  2.497        1.002      9  2.770180

[90982 rows x 5 columns]

### Function Implementation

In [ ]:
# Mampu menerapkan function pada sebuah kasus
# functian len pada dataset
len(d)
# memiliki panjang sebanyak 90982

90982

In [ ]:
pd.set_option('display.max.columns',None)

In [ ]:
d.head(5) # menampikan 5 pengukuran teratas

time(s)    V_m  Current (A)  Cycle       Cap
0        0  4.098        0.997      1  2.840817
1        1  4.096        0.998      1  2.840817
2        2  4.094        0.999      1  2.840817
3        3  4.093        0.998      1  2.840817
4        4  4.092        0.999      1  2.840817

In [ ]:
d.tail (5) # Lima pengukuran terbawah

time(s)    V_m  Current (A)  Cycle      Cap
90977    90977  2.506        1.002      9  2.77018
90978    90978  2.504        1.002      9  2.77018
90979    90979  2.502        1.002      9  2.77018
90980    90980  2.500        1.002      9  2.77018
90981    90981  2.497        1.002      9  2.77018

In [ ]:
d.shape

(90982, 5)

Berdasarkan hasil obervasi dengan d.tail, bahwa pengukuran terbawah ada di kolom 90981 dan
dengan d.head pengukuran teratas pada kolom 1. Untuk jumah keseluruhan ada 90982 baris dengan 5 kolom

In [ ]:
d.sample() # menampilkan sampel acak data set, umum nya hanya keluar satu baris

time(s)    V_m  Current (A)  Cycle       Cap
49639    49639  3.231          1.0      5  2.804878

In [ ]:
d.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 90982 entries, 0 to 90981
Data columns (total 5 columns):
 #   Column       Non-Null Count  Dtype  
---  ------       --------------  -----  
 0   time(s)      90982 non-null  int64  
 1   V_m          90982 non-null  float64
 2   Current (A)  90982 non-null  float64
 3   Cycle        90982 non-null  int64  
 4   Cap          90982 non-null  float64
dtypes: float64(3), int64(2)
memory usage: 3.5 MB


RangeIndex memiliki jumlah entri data sebanyak 90982, dimana setiap Non-nul count memiliki nilai yang sama dengan entri atau bisa di sebut tidak mengandung missing value, dan dtype dataset berupa int sama float.

In [92]:
d.describe()

time(s)           V_m   Current (A)         Cycle           Cap
count  90982.000000  90982.000000  90982.000000  90982.000000  90982.000000
mean   45490.500000      3.607666      0.999264      4.978699      2.806214
std    26264.385433      0.322777      0.001997      2.582135      0.023283
min        0.000000      2.497000      0.993000      1.000000      2.770180
25%    22745.250000      3.422000      0.997000      3.000000      2.786860
50%    45490.500000      3.636000      0.999000      5.000000      2.804878
75%    68235.750000      3.868000      1.001000      7.000000      2.823520
max    90981.000000      4.114000      1.003000      9.000000      2.840817

Misalkan interprtasi kolom v_m 
Kolom v_m memiliki tegangan yang terukur tertinggi adalah 4.11400 dan terendah adalah 2.497000. nilai rata-rata tegangan yang terukur adalah 3.607666 hampir mendekati nilai tengah sebesar 3.636000  yang dapat di artikan datanya memilki distribusi yang baik.

### Looping Implementation

In [47]:
d['V_m'].unique()

array([4.098, 4.096, 4.094, ..., 2.619, 2.552, 2.497])

In [49]:
d['V_m'].value_counts().head(3)

4.001    252
3.991    232
3.997    228
Name: V_m, dtype: int64

In [50]:
d['Cycle'].value_counts().head(3)

1    10237
2    10206
3    10175
Name: Cycle, dtype: int64

In [ ]:
# Mampu menerapkan looping for dalam suatu kasus
# kolom V_m -> d.iloc[:,3]
# kolom Cycle -> d.iloc[:,6]

hasil = []

for index in range(0,len(d)):
    selisih = d.iloc[index, 3] - d.iloc[index,6]
    hasil.append(selisih)
hasil

### Conditional If

In [ ]:
# Mampu menerapkan konsep conditional if dalam suatu kasus

In [94]:
tegangan = 2.0000
if (nilai > 3.0000):
    print("Memiliki ketegangan yang terukur")
    else :
        print("Memiliki ketegngan yang teukur di bawah")

SyntaxError: invalid syntax (1648826298.py, line 4)

### Pandas Query Implementation

In [76]:
# Mampu mengimplementasikan pandas query dalam pengolahan data
d

time(s)    V_m  Current (A)  Cycle       Cap
0            0  4.098        0.997      1  2.840817
1            1  4.096        0.998      1  2.840817
2            2  4.094        0.999      1  2.840817
3            3  4.093        0.998      1  2.840817
4            4  4.092        0.999      1  2.840817
...        ...    ...          ...    ...       ...
90977    90977  2.506        1.002      9  2.770180
90978    90978  2.504        1.002      9  2.770180
90979    90979  2.502        1.002      9  2.770180
90980    90980  2.500        1.002      9  2.770180
90981    90981  2.497        1.002      9  2.770180

[90982 rows x 5 columns]

In [77]:
tegangan = d[d['V_m']> 3.000]

In [78]:
tegangan

time(s)    V_m  Current (A)  Cycle       Cap
0            0  4.098        0.997      1  2.840817
1            1  4.096        0.998      1  2.840817
2            2  4.094        0.999      1  2.840817
3            3  4.093        0.998      1  2.840817
4            4  4.092        0.999      1  2.840817
...        ...    ...          ...    ...       ...
90480    90480  3.002        0.998      9  2.770180
90481    90481  3.002        0.999      9  2.770180
90482    90482  3.002        1.000      9  2.770180
90483    90483  3.001        1.001      9  2.770180
90484    90484  3.001        1.003      9  2.770180

[86013 rows x 5 columns]

In [79]:
tegangan.shape

(86013, 5)

Data frame tegangan yang terukur di atas 3.000 memiliki 5 columns

In [81]:
d[
    (d['V_m'] > 3.000)&
    (d['Cycle'] ==1)
    ]

time(s)    V_m  Current (A)  Cycle       Cap
0           0  4.098        0.997      1  2.840817
1           1  4.096        0.998      1  2.840817
2           2  4.094        0.999      1  2.840817
3           3  4.093        0.998      1  2.840817
4           4  4.092        0.999      1  2.840817
...       ...    ...          ...    ...       ...
9601     9601  3.002        0.997      1  2.840817
9602     9602  3.002        0.998      1  2.840817
9603     9603  3.001        0.997      1  2.840817
9604     9604  3.001        0.998      1  2.840817
9605     9605  3.001        0.998      1  2.840817

[9606 rows x 5 columns]

Berdasarkan hasil observasi data melalui query di temukan bahwa untuk kolom tegangan yang terukur waktu sebesar > 3.000 yang memiliki =1 nomor siklus charge-disharge batrei sebanyak 9606 baris. Bisa juga terlihat dari d.head() bahwa tegangan > 3.0000 memiliki nomor siklus 1.